File: examples/src/rpi_gpio3.c
#include <scriptml/gpio.h>
#include <test_print.h>
#include <unistd.h>
#include <coroutines.h>

void f(sml_gpio_handle_t hled, int *stop)
{
	while(!*stop)
	{
		sml_gpio_set_level( hled, 1 );
		sleep(1);
		sml_gpio_set_level( hled, 0 );
		sleep(1);
	}
}

int main()
{
	sml_gpio_handle_t hled = sml_gpio_open("17");
	sml_gpio_set_direction( hled, SML_GPIO_OUT );
	
	sml_gpio_handle_t hbutton = sml_gpio_open("20");
	sml_gpio_set_direction( hbutton, SML_GPIO_IN );
	
	int stop = 0;
	sml_coroutine_handle_t cid = sml_coroutine_create_2(f, hled, &stop);
	
	sml_coroutine_wait( sml_gpio_get_level( hbutton ) );
		
	printstr("button pressed");

	stop = 1;
	sml_coroutine_join(cid);

	printstr("done");

	sml_gpio_close("17");
	sml_gpio_close("20");
		
	return 0;	
}

File: examples/rpi/rpi_gpio3.py
import gpio
import time
import coroutines

def f(hled, stop):
	while not stop[0]:
		hled.set_level( 1 )
		time.sleep(1)
		hled.set_level( 0 )
		time.sleep(1)

hled = gpio.device("17");
hled.set_direction( gpio.OUT )

hbutton = gpio.device("20");
hbutton.set_direction( gpio.IN )

stop = [0]
led_coroutine = coroutines.coroutine(f, [hled, stop])

coroutines.wait( lambda: hbutton.get_level() )
	
print "button pressed"

stop[0] = 1
led_coroutine.join()

print "done"

del hled
del hbutton	
	
